### ISSUE TITLE
Should the verifier stop issuer key discovery if they already got one that worked for them?

### ISSUE THREAD

**Issue by awoie** [oauth-wg/oauth-sd-jwt-vc#281]
If the verifier supports multiple different ways of issuer key discovery, the spec currently suggests that they would need to perform all of these different ways to validate the public key. Perhaps this is a good thing, but perhaps this is an unnecessary thing? Especially if the verifier supports HTTPS and X.509, should they do both?

**Comment by bc-pi** [comment]
In my thinking, one good/working key validation/discovery is all that's needed. But, looking again, I can see how https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-08.html#section-3.5 could be read as saying to always* do both.

So I would say it's unnecessary and we need to clarify or qualify that text.

* if as a verifier you support both

**Comment by bc-pi** [comment]
Wondering, assuming really, that this came about from this https://mailarchive.ietf.org/arch/msg/oauth/oRhW0bxoDrel1mnDJR_rokHq85I/

**Comment by awoie** [comment]
> Wondering, assuming really, that this came about from this https://mailarchive.ietf.org/arch/msg/oauth/oRhW0bxoDrel1mnDJR_rokHq85I/

Indeed, it was inspired. I was supposed to send to the mailing list but then only replied individually.

**Comment by alenhorvat** [comment]
Signature validation key discovery and resolution mechanism is in most specifications defined by the signature profile and  in the VC it must be unambiguous which mechanism should be used. If the mechanism is not unique (e.g., there are 3 ways to resolve the keys) and the issuer is supporting only one, two other mechanisms could be misused by an attacker

**Comment by babisRoutis** [comment]
Hi @awoie , @bc-pi 

I was wondering the same. Have implemented this as follows:

| Supported Method (Wallet/RP) | `iss` | Method Used                                  | 
|------------------------------|-------|----------------------------------------------|
| issuer metadata              | HTTPS | issuer metadata, using `kid` if presented    |
| `x5c`                        | HTTPS | `x5c` (SAN DNS or SAN URI), if `x5c` present |
| x5c & issuer metadata        | HTTPS | `x5c` if there is `x5c`, metadata otherwise  |
| `did` (outside of sd-jwt-vc scope) | DID| `did`, using optionally `kid` (DID URL) if present|


In all cases, at most one method is used, with no fallback on error.

Given that if `x5c`method is eligible, then issuer metadata is eligible,  I think that the former (`x5c`) must take precedence, if verifier/wallet support both.

**Comment by bc-pi** [comment]
> Given that if x5cmethod is eligible, then issuer metadata is eligible, I think that the former (x5c) must take precedence, if verifier/wallet support both.

I think that's very reasonable @babisRoutis but the text in the draft needs some work. And there was some support/suggestion in yesterday's interim around having some explicit signal.

**Comment by babisRoutis** [comment]
> > Given that if x5cmethod is eligible, then issuer metadata is eligible, I think that the former (x5c) must take precedence, if verifier/wallet support both.
> 
> I think that's very reasonable @babisRoutis but the text in the draft needs some work and I'm sure could reasonably be interpreted differently. And there was some support/suggestion in yesterday's interim around having some explicit signal. Which probably/might have some relation to the concerns here https://mailarchive.ietf.org/arch/msg/oauth/ho6D8dc_7GQgJxwycS7khTM9_qw/. Some more thinking and/or work is needed in this area.

@bc-pi Thanks for sharing this. 
Hope that this will not end up with `issuer_id_scheme` :smile:

**Comment by awoie** [comment]
Ecosystems will likely define acceptable resolution methods. Therefore, this specification should not mandate any particular method. While I believe it does not, this should be explicitly clarified. Instead, the verifier (or ecosystem) should have the discretion to choose a method, as long as the necessary information to obtain the public key is present in the credential.

To further clarify, we could specify that both x5c and iss are optional, allowing an issuer to provide only the x5c header for key resolution. In the case of x5c, I assume the issuer is read from the X.509 certificate anyway, meaning the iss value should be ignored.

**Comment by babisRoutis** [comment]
Dear @awoie 

What you propose, IMO, effectively will remove one of the strongest advantages of using sd-jwt-vc format. That is,  the ability to "discover" the issuer's pub key using one of the predefined options and streamline the validation of SD-JWT-VC.

I believe that ultimately this will reduce interoperability.

PS1: Interoperability events against rapidly changing drafts sounds like an oxymoron 😄

**Comment by awoie** [comment]
> Dear [@awoie](https://github.com/awoie)
> 
> What you propose, IMO, effectively will remove one of the strongest advantages of using sd-jwt-vc format. That is, the ability to "discover" the issuer's pub key using one of the predefined options and streamline the validation of SD-JWT-VC.
> 
> I believe that ultimately this will reduce interoperability.
> 
> PS1: Interoperability events against rapidly changing drafts sounds like an oxymoron 😄

IMO, this would increase interoperability by having less options but I can see that is a bigger topic.

However, to solve this particular issue, I still believe the following:

The Verifier is not required to use more than one key resolution option even if the Verifier can. It is at the discretion of the Verifier which option is acceptable and used.

**Comment by alenhorvat** [comment]
From experience, the signature profile should clearly define one key resolution approach (I consider embedded and referenced keys in same format and same verification logic as one method) .

**Comment by awoie** [comment]
> From experience, the signature profile should clearly define one key resolution approach (I consider embedded and referenced keys in same format and same verification logic as one method) .

It would be also my preference if there is no ambiguity on the key resolution approach.

**Comment by babisRoutis** [comment]
> > Dear [@awoie](https://github.com/awoie)
> > What you propose, IMO, effectively will remove one of the strongest advantages of using sd-jwt-vc format. That is, the ability to "discover" the issuer's pub key using one of the predefined options and streamline the validation of SD-JWT-VC.
> > I believe that ultimately this will reduce interoperability.
> > PS1: Interoperability events against rapidly changing drafts sounds like an oxymoron 😄
> 
> IMO, this would increase interoperability by having less options but I can see that is a bigger topic.
> 
> However, to solve this particular issue, I still believe the following:
> 
> The Verifier is not required to use more than one key resolution option even if the Verifier can. It is at the discretion of the Verifier which option is acceptable and used.

Weekend mode on 
Perhaps, then `client_id_scheme` can also be removed and let ecosystems define the rules.
😄

**Comment by peacekeeper** [comment]
> It would be also my preference if there is no ambiguity on the key resolution approach.

This is why DIDs include the key resolution approach (aka "method") as part of the identifier itself, to avoid the kind of ambiguity discussed here, while still allowing extensibility.
